src/runtime/virtcontainers/fs_share_linux.go

```diff
@@ -507,3 +557,225 @@ func (f *FilesystemShare) UnshareRootFilesystem(ctx context.Context, c *Containe
 	return nil
 
 }
+
+func (f *FilesystemShare) watchDir(source string) error {
+
+	// Add a watcher for the configmap, secrets, projected-volumes and downwar-api directories
+	// /var/lib/kubelet/pods/<uid>/volumes/{kubernetes.io~configmap, kubernetes.io~secrets, kubernetes.io~downward-api, kubernetes.io~projected-volume}
+
+	// Note: From fsNotify docs - https://pkg.go.dev/github.com/fsnotify/fsnotify
+	// Watching individual files (rather than directories) is generally not
+	// recommended as many tools update files atomically. Instead of "just"
+	// writing to the file a temporary file will be written to first, and if
+	// successful the temporary file is moved to to destination removing the
+	// original, or some variant thereof. The watcher on the original file is
+	// now lost, as it no longer exists.
+	// Instead, watch the parent directory and use Event.Name to filter out files
+	// you're not interested in.
+
+	// Also fsNotify doesn't add watcher recursively. So we need to walk the root directory and add the required watches
+
+	f.Logger().Infof("watchDir: Add fsnotify watcher for dir (%s)", source)
+	watchList := f.watcher.WatchList()
+	if len(watchList) == 0 {
+		f.Logger().Infof("watchDir: Adding fsnotify watcher for dir (%s)", source)
+		err := f.watcher.Add(source)
+		if err != nil {
+			f.Logger().Info("watchDir: Failed to add watcher to list")
+			return err
+		}
+	}
+
+	for _, v := range watchList {
+		if v != source {
```

Comment:
Hey Pradipta - apologies for a probably dumb question, but it might help my understanding of how this is properly working.
My assumption is that `for _, v := range watchList {` will go through every source added to the watcher and so if you already had [a,b,c] added to the watchlist and called watchDir(c) the logic would be to compare each of the watched files against source e.g.
- a != c, so let's add it which would return an error as it is already watched
Have I missed something important here in the flow? The alternative I see is to loop through the watchList looking for source and then it is isn't found (after the loop) to add it?

Reply:
We didn't face any issues since there were no duplicate directories so the condition didn't occur. But I understand what you have pointed out.  Thanks for pointing out.

Reply:
Updated the code. Thanks @stevenhorsman

Reply:
Thanks! Looks good